services/data-info/src/data_info/services/metadata.clj

```diff
@@ -36,13 +37,170 @@
+   and have write permissions on the path. The path must exist. The
+   avu-map parameter must be a list of objects in this format:
+   {
+      :attr attr-string
+      :value value-string
+      :unit unit-string
+   }
+
+   Pass :system true to ignore restrictions on AVUs which may be added."
+  [user data-id avu-maps & {:keys [system] :or {system false}}]
+  (with-jargon (cfg/jargon-cfg) [cm]
+    (validators/user-exists cm user)
+    (let [path (ft/rm-last-slash (get-readable-path cm user data-id))]
+      (validators/path-writeable cm user path)
+      (when-not system (authorized-avus avu-maps))
+      (doseq [avu-map avu-maps]
+        (common-metadata-add cm path avu-map))
+      {:path path
+       :user user})))
+
+(defn admin-metadata-add
+  "Adds AVUs to path, bypassing user permission checks. See (metadata-add)
+   for the AVU map format."
+  [data-id avu-maps]
+  (metadata-add (cfg/irods-user) data-id avu-maps :system true))
+
+(defn metadata-delete
+  "Allows user to remove metadata on a path. The user must exist in iRODS
+   and have write permissions on the path. The path must exist. The
+   avu-map parameter must be a list of objects in this format:
+   {
+      :attr attr-string
+      :value value-string
+   }
+
+   Pass :system true to ignore restrictions on AVUs which may be added."
+  [user data-id avu-maps & {:keys [system] :or {system false}}]
+  (with-jargon (cfg/jargon-cfg) [cm]
+    (validators/user-exists cm user)
+    (let [path (ft/rm-last-slash (get-readable-path cm user data-id))]
+      (validators/path-writeable cm user path)
+      (when-not system (authorized-avus avu-maps))
+      (doseq [avu-map avu-maps]
+        (common-metadata-delete cm path avu-map))
+      {:path path
+       :user user})))
+
+(defn admin-metadata-delete
+  "Deletes AVUs from path, bypassing user permission checks. See (metadata-delete)
+   for the AVU map format."
+  [data-id avu-maps]
+  (metadata-delete (cfg/irods-user) data-id avu-maps :system true))
+
+(defn metadata-set
+  "Allows user to set metadata on an item with the given data-id. The user must exist in iRODS and have
+   write permissions on the data item. The 'irods-avus' parameter should be an array of AVU maps following
+   the format used for (metadata-add)."
+  [user data-id irods-avus]
+  (with-jargon (cfg/jargon-cfg) [cm]
+    (validators/user-exists cm user)
+    (let [{:keys [path type]} (uuids/path-for-uuid cm user data-id)
+          irods-avus (set (map #(select-keys % [:attr :value :unit]) irods-avus))
+          current-avus (set (list-path-metadata cm path :system false))
+          delete-irods-avus (s/difference current-avus irods-avus)]
+      (validators/path-writeable cm user path)
+      (authorized-avus irods-avus)
+
+      (doseq [del-avu delete-irods-avus]
+        (common-metadata-delete cm path del-avu))
+      (doseq [avu irods-avus]
+        (common-metadata-add cm path avu))
+
+      {:path path
+       :user user
+       :type type})))
```

Comment:
This endpoint returns the file type, so that terrain has that information in order to make a request to the metadata service.